fish-rust/src/fd_monitor.rs

```diff
@@ -0,0 +1,573 @@
+use std::os::fd::{AsRawFd, RawFd};
+use std::sync::{Arc, Mutex};
+use std::time::{Duration, Instant};
+
+use self::fd_monitor::{c_void, new_fd_event_signaller, FdEventSignaller, ItemWakeReason};
+use crate::fd_readable_set::FdReadableSet;
+use crate::fds::AutoCloseFd2;
+use crate::ffi::void_ptr;
+use crate::flog::FLOG;
+use crate::wutil::perror;
+use cxx::UniquePtr;
+
+#[cxx::bridge]
+mod fd_monitor {
+    /// Reason for waking an item
+    #[repr(u8)]
+    #[cxx_name = "item_wake_reason_t"]
+    enum ItemWakeReason {
+        /// The fd became readable (or was HUP'd)
+        Readable,
+        /// The requested timeout was hit
+        Timeout,
+        /// The item was "poked" (woken up explicitly)
+        Poke,
+    }
+
+    // Defines and exports a type shared between C++ and rust
+    struct c_void {
+        _unused: u8,
+    }
+
+    unsafe extern "C++" {
+        include!("fds.h");
+
+        /// An event signaller implemented using a file descriptor, so it can plug into
+        /// [`select()`](libc::select).
+        ///
+        /// This is like a binary semaphore. A call to [`post()`](FdEventSignaller::post) will
+        /// signal an event, making the fd readable.  Multiple calls to `post()` may be coalesced.
+        /// On Linux this uses [`eventfd()`](libc::eventfd), on other systems this uses a pipe.
+        /// [`try_consume()`](FdEventSignaller::try_consume) may be used to consume the event.
+        /// Importantly this is async signal safe. Of course it is `CLO_EXEC` as well.
+        #[rust_name = "FdEventSignaller"]
+        type fd_event_signaller_t = crate::ffi::fd_event_signaller_t;
+        // /// Mark that an event has been received.
+        // /// This may be coalesced. This retries on `EINTR`.
+        // /// TODO: This needs to be `&self` when ported to rust, otherwise it can't be used from two
+        // /// different threads.
+        // fn post(&self);
+        // /// Returns the fd to read from for notification.
+        // fn read_fd(&self) -> i32;
+        // /// If an event is signalled, consume it; otherwise return.
+        // /// This does not block. This retries on `EINTR`.
+        // fn try_consume(&self) -> bool;
+        #[rust_name = "new_fd_event_signaller"]
+        fn ffi_new_fd_event_signaller_t() -> UniquePtr<FdEventSignaller>;
+    }
+    // unsafe extern "C++" {
+    //     include!("fds.h");
+    //
+    //     /// A helper class for managing and automatically closing a file descriptor.
+    //     #[rust_name = "AutoCloseFd"]
+    //     type autoclose_fd_t; // = crate::ffi::autoclose_fd_t;
+    //     /// Closes the fd if not already closed
```

Comment:
should probably not add commented code

Reply:
Yeah I kept flipping between using the rust and cpp versions of the auto close type. I’ll remove it.